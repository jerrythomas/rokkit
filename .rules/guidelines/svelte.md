# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
