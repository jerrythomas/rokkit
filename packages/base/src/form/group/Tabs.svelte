<script>
	import TabItems from './TabItems.svelte'
	import TabItem from './TabItem.svelte'

	export let items = []
	export let fields = {}
	export let title = null
	export let allowAdd = false
	export let allowClose = false
	export let activeItem = items[0]

	function addTab() {
		items = [...items, {}]
		activeItem = items[items.length - 1]
	}

	$: console.log(activeItem)
</script>

<tab-view class="flex flex-col w-full flex-grow">
	<tabs class="flex flex-row flex-shrink-0 w-full select-none cursor-pointer">
		{#if title}
			<p>{title}</p>
		{/if}
		<TabItems {items} {fields} {allowClose} bind:activeItem on:close />
		{#if allowAdd}
			<TabItem label="+" on:click={addTab} />
		{/if}
	</tabs>
	<content class="flex flex-col flex-grow"><slot /></content>
</tab-view>
