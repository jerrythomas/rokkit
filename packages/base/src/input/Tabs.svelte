<script>
	import TabItems from './TabItems.svelte'
	import TabItem from './TabItem.svelte'

	export let items = []
	export let fields = {}
	export let allowAdd = false
	export let allowClose = false
	export let activeItem = items[0]

	function addTab() {
		items = [...items, {}]
		activeItem = items[items.length - 1]
	}
</script>

<tabs class="flex flex-row flex-grow select-none cursor-pointer">
	<TabItems {items} {fields} {allowClose} bind:activeItem on:close />
	{#if allowAdd}
		<TabItem label="+" on:click={addTab} />
	{/if}
</tabs>
