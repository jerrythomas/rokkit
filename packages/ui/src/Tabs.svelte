<script>
	import { defaultFields, defaultStateIcons, noop } from '@rokkit/core'
	import { defaultMapping } from './constants'
	import { navigator } from '@rokkit/actions'
	import Icon from './Icon.svelte'
	import Item from './Item.svelte'

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [options]
	 * @property {import('@rokkit/core').FieldMapping} [mapping]
	 * @property {any} [value]
	 * @property {boolean} [below]
	 * @property {string} [align]
	 * @property {boolean} [editable]
	 * @property {any} [icons]
	 */

	/** @type {Props} */
	let {
		class: className = '',
		options = $bindable([]),
		fields = defaultFields,
		mapping = defaultMapping,
		value = $bindable(null),
		below = false,
		align = 'left',
		editable = false,
		icons = $bindable(defaultStateIcons.action),
		onremove = noop,
		onadd = noop,
		onselect = noop
	} = $props()

	let cursor = $state([])

	function handleRemove(event) {
		if (typeof event.detail === Object) {
			event.detail[mapping.fields.isDeleted] = true
		} else {
			options = options.filter((i) => i !== event.detail)
		}

		onremove({ item: event.detail })
	}
	function handleAdd(event) {
		event.stopPropagation()
		onadd()
	}
	function handleNav(event) {
		value = event.detail.node
		cursor = event.detail.path

		onselect({ item: value, indices: cursor })
	}
	let stateIcons = $derived({ ...defaultStateIcons.action, ...icons })
	let filtered = $derived(options.filter((item) => !item[fields.isDeleted]))
</script>

<rk-tabs
	class="flex w-full {className}"
	class:is-below={below}
	class:justify-center={align === 'center'}
	class:justify-end={align === 'right'}
	tabindex="0"
	role="listbox"
	onmove={handleNav}
	onselect={handleNav}
>
	{#each filtered as item, index}
		{@const Template = mapping.getComponent(item)}
		<rk-tab>
			<Template value={item} {mapping} />
			{#if editable}
				<Icon
					name="remove"
					role="button"
					label="Delete Tab"
					size="small"
					onclick={() => handleRemove(item)}
				/>
			{/if}
		</rk-tab>
	{/each}
	{#if editable}
		<Icon name="add" role="button" label="Add Tab" size="small" on:click={handleAdd} />
	{/if}
</rk-tabs>
