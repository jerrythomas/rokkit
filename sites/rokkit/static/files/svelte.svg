<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M18.2301 11.1217C21.1087 6.47773 15.3383 1.6935 11.2372 4.37297L7.337 6.85957C5.44834 8.08961 4.62832 10.453 5.78004 12.8817C2.90038 17.2501 8.50365 22.3655 12.7729 19.6332L16.6732 17.1466C18.7952 15.7494 19.2821 13.3398 18.2301 11.1217V11.1217ZM8.46536 8.46761L12.3599 5.98242C14.9184 4.45077 17.8644 7.12896 16.8023 9.62435C16.1715 9.12318 15.4685 8.81503 14.9345 8.67627C15.1375 7.87006 14.2016 7.27202 13.5507 7.68841L9.64763 10.1651C8.68258 10.8321 9.71148 12.204 10.6653 11.6284L12.1527 10.6803C15.4905 8.79676 18.6495 13.3192 15.5448 15.5357L11.6502 18.0209C9.30336 19.5215 6.10922 17.1175 7.20785 14.379C7.79145 14.8207 8.41131 15.1456 9.07564 15.3271C8.92896 16.2531 9.86333 16.7033 10.4694 16.3234L14.3626 13.8382C15.3233 13.1178 14.292 11.814 13.3464 12.375L11.8533 13.3231C8.55362 15.2514 5.32106 10.6843 8.46536 8.46761Z" fill="black"/>
</svg>
