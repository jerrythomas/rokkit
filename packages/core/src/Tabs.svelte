<script>
	import { defaultFields } from './constants'
	import { Text } from './items'
	import { navigator } from './actions'
	import { createEventDispatcher } from 'svelte'

	const dispatch = createEventDispatcher()

	let className = ''
	export { className as class }
	export let items = []
	export let fields = {}
	export let using = {}
	export let value = null
	export let below = false
	export let align = 'left'

	let cursor = []

	function handleNav(event) {
		value = event.detail.node
		cursor = event.detail.path

		dispatch('select', { item: value, indices: cursor })
	}
	$: fields = { ...defaultFields, ...fields }
	$: using = { default: Text, ...using }
</script>

<tabs
	class="flex w-full {className}"
	class:is-below={below}
	class:justify-center={align == 'center'}
	class:justify-end={align == 'right'}
	tabindex="0"
	role="listbox"
	use:navigator={{
		items,
		fields,
		vertical: false,
		indices: cursor
	}}
	on:move={handleNav}
	on:select={handleNav}
>
	{#each items as item, index}
		{@const component = item[fields.component]
			? using[item[fields.component]] || using.default
			: using.default}

		<item
			class="flex"
			role="option"
			aria-selected={item === value}
			class:is-selected={item === value}
			data-path={index}
		>
			<svelte:component this={component} content={item} {fields} />
		</item>
	{/each}
</tabs>
