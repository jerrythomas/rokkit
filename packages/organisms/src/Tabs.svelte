<script>
	import { createEventDispatcher } from 'svelte'
	import { defaultFields, defaultStateIcons } from '@rokkit/core'
	import { navigator } from '@rokkit/actions'
	import { Icon } from '@rokkit/atoms'
	import { ItemWrapper, Item } from '@rokkit/molecules'

	const dispatch = createEventDispatcher()

	let className = ''
	export { className as class }
	export let options = []
	/** @type {import('@rokkit/core').FieldMapping} */
	export let fields = {}
	export let using
	export let value = null
	export let below = false
	export let align = 'left'
	export let editable = false
	export let icons = defaultStateIcons.action

	let cursor = []

	function handleRemove(item) {
		if (typeof item === Object) {
			item[fields.isDeleted] = true
		} else {
			options = options.filter((i) => i !== item)
		}

		dispatch('remove', { item })
	}
	function handleNav(event) {
		value = event.detail.node
		cursor = event.detail.path

		dispatch('select', { item: value, indices: cursor })
	}
	$: using = { default: Item, ...using }
	$: icons = { ...defaultStateIcons.action, ...icons }
	$: filtered = options.filter((item) => !item[fields.isDeleted])
	$: fields = { ...defaultFields, ...fields }
</script>

<tabs
	class="flex w-full {className}"
	class:is-below={below}
	class:justify-center={align == 'center'}
	class:justify-end={align == 'right'}
	tabindex="0"
	role="listbox"
	use:navigator={{
		items: options,
		fields,
		vertical: false,
		indices: cursor
	}}
	on:move={handleNav}
	on:select={handleNav}
>
	{#each filtered as item, index}
		<ItemWrapper
			value={item}
			{index}
			{fields}
			{using}
			removable={editable}
			selected={item === value}
			{icons}
			class="tab"
			on:remove={handleRemove}
		/>
	{/each}
	{#if editable}
		<Icon
			name="add"
			role="button"
			label="Add Tab"
			size="small"
			on:click={() => dispatch('add')}
		/>
	{/if}
</tabs>
