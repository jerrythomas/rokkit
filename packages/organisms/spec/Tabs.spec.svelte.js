import { describe, it, expect, beforeEach } from 'vitest'
import { render, cleanup } from '@testing-library/svelte'
import Tabs from '../src/Tabs.svelte'
import { flushSync } from 'svelte'

describe('Tabs', () => {
	beforeEach(() => cleanup())

	it('should render simple list', () => {
		const props = $state({ items: ['Item 1', 'Item 2'], class: 'custom-class' })
		const { container } = render(Tabs, props)
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		props.class = 'another-class'
		flushSync()
		expect(container).toMatchSnapshot()
	})
})
