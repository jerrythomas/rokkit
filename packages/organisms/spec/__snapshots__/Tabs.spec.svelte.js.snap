// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tabs > should render simple list 1`] = `
<div>
  <rk-tabs
    class="flex w-full custom-class"
    role="listbox"
    tabindex="0"
  >
    <!---->
     
    <!---->
  </rk-tabs>
  
</div>
`;

exports[`Tabs > should render simple list 2`] = `
<div>
  <rk-tabs
    class="flex w-full another-class"
    role="listbox"
    tabindex="0"
  >
    <!---->
     
    <!---->
  </rk-tabs>
  
</div>
`;
